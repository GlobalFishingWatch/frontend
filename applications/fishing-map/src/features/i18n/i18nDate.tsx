import React, { Fragment } from 'react'
import { DateTime, DateTimeFormatOptions } from 'luxon'
import { useTranslation } from 'react-i18next'

type Dates = {
  date: string
  format?: DateTimeFormatOptions
}

export const formatI18nDate = (date: string, locale: string, format = DateTime.DATE_MED) => {
  return `${DateTime.fromISO(date).toUTC().setLocale(locale).toLocaleString(format)}${
    format === DateTime.DATETIME_MED ? ' UTC' : ''
  }`
}

export const useI18nDate = (date: string, format = DateTime.DATE_MED) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, i18n.language, format)
}

const I18nDate = ({ date, format = DateTime.DATE_MED }: Dates) => {
  const dateFormatted = useI18nDate(date, format)
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
