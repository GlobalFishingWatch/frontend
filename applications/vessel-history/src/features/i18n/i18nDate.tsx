import React, { Fragment } from 'react'
import { DateTime, DateTimeFormatOptions } from 'luxon'
import { useTranslation } from 'react-i18next'
import { Locale } from 'types'
import i18n from './i18n'

type Dates = {
  date: string | number
  format?: DateTimeFormatOptions
}

type formatI18DateParams = { format?: DateTimeFormatOptions; locale?: Locale }

export const formatI18nDate = (
  date: string | number,
  { format = DateTime.DATE_MED, locale = i18n.language as Locale }: formatI18DateParams = {}
) => {
  const dateTimeDate = typeof date === 'number' ? DateTime.fromMillis(date) : DateTime.fromISO(date)
  return `${dateTimeDate.toUTC().setLocale(locale).toLocaleString(format)}${
    format === DateTime.DATETIME_MED ? ' UTC' : ''
  }`
}

export const useI18nDate = (date: string | number, format = DateTime.DATE_MED) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, { format, locale: i18n.language as Locale })
}

const I18nDate = ({ date, format = DateTime.DATE_MED }: Dates) => {
  const dateFormatted = useI18nDate(date, format)
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
