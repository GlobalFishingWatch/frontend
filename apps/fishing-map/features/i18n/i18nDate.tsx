import { Fragment } from 'react'
import { useTranslation } from 'react-i18next'
import type { DateTimeFormatOptions } from 'luxon'
import { DateTime } from 'luxon'

import type { SupportedDateType } from '@globalfishingwatch/data-transforms'

import type { Locale } from 'types'
import { getUTCDateTime } from 'utils/dates'

import i18n from './i18n'

type Dates = {
  date: string | number
  format?: DateTimeFormatOptions
  showUTCLabel?: boolean
}

type formatI18DateParams = {
  format?: DateTimeFormatOptions | object
  locale?: Locale
  showUTCLabel?: boolean
}

const UTC_SUFFIX = 'UTC'

export const formatI18nDate = (
  date: SupportedDateType,
  {
    format = DateTime.DATE_MED,
    locale = i18n.language as Locale,
    showUTCLabel = false,
  }: formatI18DateParams = {}
) => {
  const dateTimeDate = getUTCDateTime(date)
  return `${dateTimeDate?.setLocale(locale).toLocaleString(format)}${
    format === DateTime.DATETIME_MED || showUTCLabel ? ` ${UTC_SUFFIX}` : ''
  }`
}

export const useI18nDate = (
  date: SupportedDateType,
  format = DateTime.DATE_MED,
  showUTCLabel = false
) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, { format, locale: i18n.language as Locale, showUTCLabel })
}

const I18nDate = ({ date, format = DateTime.DATE_MED, showUTCLabel = false }: Dates) => {
  const dateFormatted = useI18nDate(date, format, showUTCLabel)
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
