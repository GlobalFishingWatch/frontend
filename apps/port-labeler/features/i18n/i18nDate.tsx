import React, { Fragment } from 'react'
import { useTranslation } from 'react-i18next'
import type { DateTimeFormatOptions } from 'luxon';
import { DateTime } from 'luxon'
import type { Locale } from 'types'

import { getUTCDateTime } from 'utils/dates'

import i18n from './i18n'

type Dates = {
  date: string
  format?: DateTimeFormatOptions
}

type formatI18DateParams = { format?: DateTimeFormatOptions; locale?: Locale }

export const formatI18nDate = (
  date: string | number,
  { format = DateTime.DATE_MED, locale = i18n.language as Locale }: formatI18DateParams = {}
) => {
  const dateTimeDate = getUTCDateTime(date)
  return `${dateTimeDate.setLocale(locale).toLocaleString(format)}${
    format === DateTime.DATETIME_MED ? ' UTC' : ''
  }`
}

export const useI18nDate = (date: string, format = DateTime.DATE_MED) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, { format, locale: i18n.language as Locale })
}

const I18nDate = ({ date, format = DateTime.DATE_MED }: Dates) => {
  const dateFormatted = useI18nDate(date, format)
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
