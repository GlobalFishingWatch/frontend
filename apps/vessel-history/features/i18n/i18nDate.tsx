import React, { Fragment } from 'react'
import { useTranslation } from 'react-i18next'
import type { DateTimeFormatOptions} from 'luxon';
import { DateTime, Zone } from 'luxon'
import type { Locale } from 'types'

import { getUTCDateTime } from 'utils/dates'

import i18n from './i18n'

type Dates = {
  date: string | number
  format?: DateTimeFormatOptions
}

type TMTDate = {
  date: string | number
  format?: DateTimeFormatOptions
  forcedTokensFormat?: string
}

type formatI18DateParams = {
  format?: DateTimeFormatOptions
  locale?: Locale
  tokensFormat?: string
}

export const formatI18nDate = (
  date: string | number,
  {
    format = DateTime.DATE_MED,
    locale = i18n.language as Locale,
    tokensFormat,
  }: formatI18DateParams = {}
) => {
  const dateTimeDate = getUTCDateTime(date).setLocale(locale)
  // console.log(
  //   (() => {
  //     const a = DateTime.fromISO('20200101').setZone('utc', { keepLocalTime: true })
  //     const b = DateTime.fromISO('2020-01-01T00:00:00', { zone: 'utc' })
  //     const c = DateTime.fromISO('2020-01-01T00:00:00')
  //     return {
  //       a: a.toISO(),
  //       b: b.toISO(),
  //       c: c.toUTC().toISO(),
  //     }
  //   })()
  // )
  // console.log(new Date().getTime() - DateTime.now().toUTC().toMillis())
  const formattedDate = tokensFormat
    ? dateTimeDate.toFormat(tokensFormat)
    : dateTimeDate.toLocaleString(format)
  return `${formattedDate}${format === DateTime.DATETIME_MED ? ' UTC' : ''}`
}

export const useI18nDate = (
  date: string | number,
  format = DateTime.DATE_MED,
  tokensFormat?: string
) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, { format, locale: i18n.language as Locale, tokensFormat })
}

const I18nDate = ({ date, format = DateTime.DATE_MED }: Dates) => {
  const dateFormatted = useI18nDate(date, format)
  return <Fragment>{dateFormatted}</Fragment>
}

export const formatI18nSpecialDate = ({ date, format = DateTime.DATE_MED }: TMTDate) => {
  let parsedDate, tokensFormat
  if (date.toString().match(/^\d{4}(00|99){1,2}$/)) {
    parsedDate = date.toString().replace(/(\d{4})(00|99){1,2}$/, '$1-01-01')
    tokensFormat = 'yyyy'
  } else if (date.toString().match(/^\d{6}(00|99)$/)) {
    parsedDate = date.toString().replace(/(\d{4})(\d{2})(00|99)$/, '$1-$2-01')
    tokensFormat = 'LLL, yyyy'
  } else if (date.toString().match(/^\d{4}-\d{2}-9{2}$/)) {
    // Converts 2016-02-99 dates to 2016-02-01 and formats it as Feb, 2016
    // TODO: Confirm with the source if this interpretation is correct
    parsedDate = date.toString().replace(/(\d{4})-(\d{2})-(9{2})$/, '$1-$2-01')
    tokensFormat = 'LLL, yyyy'
  } else if (date.toString().match(/^\d{8}$/)) {
    parsedDate = date.toString().replace(/(\d{4})(\d{2})(\d{2})$/, '$1-$2-$3')
  }
  return formatI18nDate(parsedDate ?? date, {
    format,
    locale: i18n.language as Locale,
    tokensFormat: tokensFormat,
  })
}

export const I18nSpecialDate = ({ date, format = DateTime.DATE_MED }: TMTDate) => {
  const dateFormatted = formatI18nSpecialDate({ date, format })
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
