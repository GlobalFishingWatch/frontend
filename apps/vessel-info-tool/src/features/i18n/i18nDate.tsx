import { Fragment } from 'react'
import { useTranslation } from 'react-i18next'
import type { DateTimeFormatOptions } from 'luxon'
import { DateTime } from 'luxon'

import type { Locale } from '@globalfishingwatch/api-types'
import { getUTCDateTime, type SupportedDateType } from '@globalfishingwatch/data-transforms'

import i18n from './i18n'

type Dates = {
  date: string | number
  format?: DateTimeFormatOptions
  showUTCLabel?: boolean
}

type formatI18DateParams = {
  format?: DateTimeFormatOptions | object
  locale?: Locale
  showUTCLabel?: boolean
}

const UTC_SUFFIX = 'UTC'

export const formatI18nDate = (
  date: SupportedDateType,
  {
    format = DateTime.DATE_MED,
    locale = i18n.language as Locale,
    showUTCLabel,
  }: formatI18DateParams = {}
) => {
  const dateTimeDate = getUTCDateTime(date)
  let utcSuffix = ''
  if ((format === DateTime.DATETIME_MED && showUTCLabel === undefined) || showUTCLabel) {
    utcSuffix = ` ${UTC_SUFFIX}`
  }
  return `${dateTimeDate?.setLocale(locale).toLocaleString(format)}${utcSuffix}`
}

export const useI18nDate = (
  date: SupportedDateType,
  format = DateTime.DATE_MED,
  showUTCLabel?: boolean
) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, { format, locale: i18n.language as Locale, showUTCLabel })
}

const I18nDate = ({ date, format = DateTime.DATE_MED, showUTCLabel }: Dates) => {
  const dateFormatted = useI18nDate(date, format, showUTCLabel)
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
