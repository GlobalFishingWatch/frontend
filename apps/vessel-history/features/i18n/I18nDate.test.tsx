import { render } from '@testing-library/react'

import { I18nSpecialDate } from './i18nDate'

jest.mock('features/i18n/i18n')

describe('<I18nSpecialDate />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('formats TMT date only with year', () => {
    const component = render(<I18nSpecialDate date={19800000} />)
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('formats TMT date with month and year', () => {
    const component = render(<I18nSpecialDate date={19801100} />)
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('formats TMT date with day, month and year', () => {
    const component = render(<I18nSpecialDate date={19801121} />)
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('formats date from ISO format', () => {
    const component = render(<I18nSpecialDate date={'2021-01-01T03:59:51Z'} />)
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('formats date from timestamp', () => {
    const component = render(<I18nSpecialDate date={1378094391000} />)
    expect(component.asFragment()).toMatchSnapshot()
  })
})
