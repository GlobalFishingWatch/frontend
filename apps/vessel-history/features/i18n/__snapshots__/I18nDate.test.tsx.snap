// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<I18nSpecialDate /> formats TMT date only with year 1`] = `
<DocumentFragment>
  1980
</DocumentFragment>
`;

exports[`<I18nSpecialDate /> formats TMT date with day, month and year 1`] = `
<DocumentFragment>
  Nov 21, 1980
</DocumentFragment>
`;

exports[`<I18nSpecialDate /> formats TMT date with month and year 1`] = `
<DocumentFragment>
  Nov, 1980
</DocumentFragment>
`;

exports[`<I18nSpecialDate /> formats date from ISO format 1`] = `
<DocumentFragment>
  Jan 1, 2021
</DocumentFragment>
`;

exports[`<I18nSpecialDate /> formats date from timestamp 1`] = `
<DocumentFragment>
  Sep 2, 2013
</DocumentFragment>
`;
